src/__tests__/typeChecker.test.ts

```diff
@@ -0,0 +1,311 @@
+"accumulate: ((T67, T21) -> T21, T21, [T67, List<T67>]) -> T21
+xs: [number, List<number>]
+ys: [boolean, List<boolean>]"
+`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('Will not work if used in a monomorphic manner', () => {
+    const code = `
+      function accumulate(op, init, xs) {
+        return is_null(xs) ? init : op(head(xs), accumulate(op, init, tail(xs)));
+      }
+      const xs = pair(1, pair(2, null));
+      const ys = pair(true, pair(true, null));
+      const a = accumulate((x,y)=>x+y,0,xs);
+      const b = accumulate((x,y)=>x||y,0,ys);
+    `
+
+    const [program, errors] = typeCheck(parse(code, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+"accumulate: ((number, number) -> number, number, [number, List<number>]) -> number
+xs: [number, List<number>]
+ys: [boolean, List<boolean>]
+a: number
+b: number"
+`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+"Line 8: Types do not unify: number vs boolean
+Line 8: Types do not unify: boolean vs number"
+`)
+  })
+})
+
+describe('type checking functions', () => {
+  it('happy paths for recursive functions', () => {
+    const code1 = `
+      function append(xs, ys) {
+        return is_null(xs) ? ys : pair(head(xs), append(tail(xs), ys));
+      }
+    `
+    const [program, errors] = typeCheck(parse(code1, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(
+      `"append: ([T29, List<T29>], [T29, List<T29>]) -> [T29, List<T29>]"`
+    )
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('unhappy paths for recursive functions', () => {
+    const code = `
+      function foo(f) {
+        return foo;
+      }
+    `
+    const [program, errors] = typeCheck(parse(code, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`"foo: Couldn't infer type"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+"Line 1: Contains cyclic reference to itself, where the type being bound to is a function type
+Line 2: Error: Contains cyclic reference to itself, where the type being bound to is a function type
+Line 2: Error: Contains cyclic reference to itself, where the type being bound to is a function type
+Line 3: Error: Contains cyclic reference to itself, where the type being bound to is a function type
+Line 3: Error: Contains cyclic reference to itself, where the type being bound to is a function type"
+`)
+  })
+})
+
+describe('type checking pairs', () => {
+  it('wrapping pair functions', () => {
+    const code = `
+function foo(x, y) {
+  return pair(x, y);
+}
+    `
+    const [program, errors] = typeCheck(parse(code, 2))
+    expect(topLevelTypesToString(program!)).toMatchInlineSnapshot(`"foo: (T3, T4) -> [T3, T4]"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('happy paths for pair functions', () => {
+    const code = `
+const x = pair(3, 4);
+head(x) + 56;
```

Comment:
should probably move 184 down so that the type of foo can be inferred.

On another note, why does y have a correct type when foo doesn't?

Reply:
I think it has got to do with the validity checker being a bit too restrictive. If it was a type error bug in our checker it would have raised a error, but we didnt raise any such bugs. Reordering the statements seems to make the issue go away so I will just do that.

Reply:
Fixed it by reordering the statements.